test: add tests for C99 edge cases

anandkaranubc opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

- Completes the TODO on confirming the alignment of `math/base/special/cabs` with C99 conventions

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
chore: improve implementation and tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/cabs/lib/main.js:
```diff
@@ -40,7 +43,12 @@ var imag = require( '@stdlib/complex/float64/imag' );
 */
 function cabs( z ) {
-	// TODO: consider whether to use C99 rules for special cases involving infinities and nans (see https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Objects/complexobject.c#L191)
+	if ( isnan( real( z ) ) || isnan( imag( z ) ) ) {
+		return NaN;
+	}
+	if ( isInfinite( real( z ) ) || isInfinite( imag( z ) ) ) {
+		return PINF;
+	}
```
This essentially is not "explicitly" needed, as all this is also being checked by `hypot`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/cabs/test/test.native.js:
```diff
@@ -80,13 +82,43 @@ tape( 'if either the real or imaginary component is `NaN`, the function returns
+	t.end();
+});
+
+tape( 'if either the real or imaginary component is `+Infinity`, the function returns `+Infinity`', opts, function test( t ) {
```
However, these tests could help improve the test coverage.

anandkaranubc commented:
Also, based on the TODO comment referring to SciPy's [implementation](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Objects/complexobject.c#L191), I am still not sure why they perform an infinity check before a NaN check. They also specify that it's according to the C99 rules. However, testing with those values shows something different—if any of the arguments is NaN, the result is NaN.

[Test results](https://coliru.stacked-crooked.com/view?id=411fca582c7f59a9)

No special comment has been made in the reference C99 [documentation](https://en.cppreference.com/w/c/numeric/complex/cabs), and I couldn’t find anything in the [ISO/IEC 9899:1999](https://www.dii.uchile.cl/~daespino/files/Iso_C_1999_definition.pdf) that justifies this either.

cc: @kgryte

kgryte commented:
@anandkaranubc See also https://data-apis.org/array-api/latest/API_specification/generated/array_api.abs.html#array_api.abs. The rule to check for infinity before NaN can be found in C99 hypot: https://en.cppreference.com/w/c/numeric/math/hypot.

kgryte commented:
It's also possible that our `hypot` is not IEEE 754 compliant. We should double check, as our `hypot` does not match C99 atm.

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into fix/cabs

anandkaranubc committed:
fix: update implementation to follow the C99 standard

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: remove extra linebreak

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM. Thanks, @anandkaranubc!
